internal/subscriber/retry/delayconsumer_test.go

```diff
@@ -81,6 +91,118 @@ func TestDelayConsumer_Run(t *testing.T) {
 	dc.Run(ctx)
 }
 
+func TestDelayConsumer_Run_Consume(t *testing.T) {
+	ctx, subCtx, cancel, store, handler, consumer, cache := setupDC(t)
+	subs := getDummySubscriptionModel()
+	subscriberID := "subscriber-id"
+	errCh := make(chan error)
+	dc, err := NewDelayConsumer(subCtx, subscriberID, "t1", subs, store, handler, cache, errCh)
+	assert.NotNil(t, dc)
+	assert.Nil(t, err)
+	assert.NotNil(t, dc.LogFields())
+	msgs := make([]messagebroker.ReceivedMessage, 0)
+	msg1 := getDummyBrokerMessage()
+	msg1.NextDeliveryTime = time.Now().Add(time.Nanosecond * 1000)
+	msg2 := getDummyBrokerMessage()
+	msg2.NextDeliveryTime = time.Now().Add(time.Second * 2)
+	msgs = append(msgs, msg1, msg2)
+	type args struct {
+		messages []messagebroker.ReceivedMessage
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name:    "Test Delay Consumer one message counsume out of 2",
+			wantErr: false,
+			args: args{
+				messages: msgs,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			count := 0
+			consumer.EXPECT().ReceiveMessages(gomock.Any(), gomock.Any()).DoAndReturn(
+				func(arg0 context.Context, arg1 messagebroker.GetMessagesFromTopicRequest) (*messagebroker.GetMessagesFromTopicResponse, error) {
+					if count == 0 {
+						count++
+						return &messagebroker.GetMessagesFromTopicResponse{Messages: tt.args.messages}, nil
+					} else {
+						return &messagebroker.GetMessagesFromTopicResponse{}, nil
+					}
+				}).AnyTimes()
+			go dc.Run(ctx)
+			select {
+			case err := <-errCh:
+				if !tt.wantErr {
+					t.Errorf("Got Error : %v", err)
+				}
+			case <-time.NewTicker(time.Millisecond * 500).C:
+				assert.Equal(t, len(dc.cachedMsgs), 1)
```

Comment:
this is quite specific to how dc works..shouldn't we figure out some way to see if message is consumed by dc

Reply:
there are 2 messages and one message is consumed since there is delay of only 1000 nanosecond , and other one is not consumed since delay is of 2 sec, thats why we have cached messages length = 1 , making sure message being consumed by DC, i am not able to find a way to determine if a message is consumed by the Defined DC or not.

Reply:
cachedMsgs will always consume message in batch irrespective of nextDeliveryTime, but will pause the consumer if the consumed message should not be processed. 
You can write test case for CanProcessMessage() func i think

Reply:
While processing the message it checking if it can process the message or not , if it process the message then it is being removed from cached Msgs that why added only one message will be present in the cached messages

Reply:
can we then compare the data in cachedMsgs to the one we have in the input..?

Reply:
compared the cachedMsgs with input